genesis/genesis.go

```diff
@@ -0,0 +1,146 @@
+package genesis
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/NethermindEth/juno/adapters/sn2core"
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/starknet"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/NethermindEth/juno/validator"
+	"github.com/NethermindEth/juno/vm"
+)
+
+type GenesisConfig struct {
+	ChainID string `json:"chain_id" validate:"required"`
+
+	Classes []string `json:"classes"` // []path-to-class.json
+
+	Contracts map[string]GenesisContractData `json:"contracts"` // address -> {classHash, constructorArgs}
+
+	FunctionCalls []FunctionCall `json:"function_calls"` // list of functionCalls to Call()
+}
+
+type GenesisContractData struct {
+	ClassHash       felt.Felt   `json:"class_hash"`
+	ConstructorArgs []felt.Felt `json:"constructor_args"`
+}
+type FunctionCall struct {
+	ContractAddress    felt.Felt   `json:"contract_address"`
+	EntryPointSelector felt.Felt   `json:"entry_point_selector"`
+	Calldata           []felt.Felt `json:"calldata"`
+}
+
+func (g *GenesisConfig) Validate() error {
+	validate := validator.Validator()
+	return validate.Struct(g)
+}
+
+// GenesisStateDiff builds the genesis stateDiff given the genesis-config data.
+func GenesisStateDiff(
+	genesisConfig GenesisConfig,
+	chain *blockchain.Blockchain,
+	vmm vm.VM,
+	network *utils.Network,
+	log utils.Logger,
+) (*core.StateDiff, map[felt.Felt]core.Class, error) {
+	blockTimestamp := uint64(time.Now().Unix())
+
+	newClasses, err := loadClasses(genesisConfig.Classes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pendingReader, closer, err := chain.PendingState()
```

Comment:
We can get rid of the `chain` dependency here. See how we create an empty state in `blockchain`

```
		txn := db.NewMemTransaction()
		emptyState := core.NewState(txn)
		return emptyState, txn.Discard, nil
```